src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderStates/OrderStateType.php

```diff
@@ -0,0 +1,205 @@
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\Form\Admin\Configure\ShopParameters\OrderStates;
+
+use PrestaShop\PrestaShop\Adapter\Configuration;
+use PrestaShop\PrestaShop\Adapter\MailTemplate\MailTemplateDataProvider;
+use PrestaShop\PrestaShop\Core\ConstraintValidator\Constraints\DefaultLanguage;
+use PrestaShopBundle\Form\Admin\Type\ColorPickerType;
+use PrestaShopBundle\Form\Admin\Type\TranslatableChoiceType;
+use PrestaShopBundle\Form\Admin\Type\TranslatableType;
+use PrestaShopBundle\Form\Admin\Type\TranslatorAwareType;
+use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Routing\Router;
+use Symfony\Component\Translation\TranslatorInterface;
+use Symfony\Component\Validator\Constraints\Regex;
+
+/**
+ * Type is used to created form for order state add/edit actions
+ */
+class OrderStateType extends TranslatorAwareType
+{
+    /**
+     * @var array
+     */
+    private $templates;
+    /**
+     * @var Router
+     */
+    private $routing;
+    /**
+     * @var array
+     */
+    private $templateAttributes;
+
+    public function __construct(
+        TranslatorInterface $translator,
+        array $locales,
+        MailTemplateDataProvider $mailTemplateDataProvider,
+        Router $routing
+    ) {
+        parent::__construct($translator, $locales);
+        $this->routing = $routing;
+
+        foreach ($mailTemplateDataProvider->getTemplates() as $languageId => $languageMailTemplate) {
+            $this->templates[$languageId] = [];
+            $this->templateAttributes[$languageId] = [];
+            foreach ($languageMailTemplate as $mailTemplate) {
+                if (array_key_exists('id', $mailTemplate) && array_key_exists('name', $mailTemplate)) {
+                    $this->templates[$languageId][$mailTemplate['name']] = $mailTemplate['id'];
+
+                    $this->templateAttributes[$languageId][$mailTemplate['name']] = [];
+
+                    $this->templateAttributes[$languageId][$mailTemplate['name']]['data-preview'] = $this->routing->generate(
+                        'admin_mail_theme_preview_layout',
+                        [
+                            'locale' => $mailTemplate['language_code'],
+                            'theme' => Configuration::get('PS_MAIL_THEME', 'modern'),
```

Comment:
This Configuration can be in a variable outside the two foreach :thinking:

Reply:
done